src/mlpack/methods/reinforcement_learning/environment/frozen_lake.hpp

```diff
@@ -0,0 +1,414 @@
+      nCols(other.nCols),
+      curRow(other.curRow),
+      curCol(other.curCol)
+    {
+      // Nothing to do here.
+    }
+
+    //! Operator= copy constructor.
+    State& operator=(State const& other)
+    {
+      if (this != &other)
+      {
+        nRows = other.nRows;
+        nCols = other.nCols;
+        curRow = other.curRow;
+        curCol = other.curCol;
+      }
+      return *this;
+    }
+
+    //! Operator= move constructor.
+    State& operator=(State&& other)
+    {
+      if (this != &other)
+      {
+        nRows = other.nRows;
+        nCols = other.nCols;
+        curRow = other.curRow;
+        curCol = other.curCol;
+      }
+      return *this;
+    }
+
+    //! Get nRows.
+    size_t NumRows() const {return nRows;}
+
+    //! Get nCol.
+    size_t NumCols() const {return nCols;}
+
+    //! Get current row.
+    size_t CurRow() const {return curRow;}
+
+    //! Get current col.
+    size_t CurCol() const {return curCol;}
+
+    //! Set current row.
+    size_t& CurRow() {return curRow;}
+
+    //! Set current col.
+    size_t& CurCol() {return curCol;}
+
+    //! Get state representation.
+    size_t GetState() const {return nCols * curRow + curCol;}
+
+   private:
+    size_t nRows;
+    size_t nCols;
+    size_t curRow;
+    size_t curCol;
+  };
+
+  /**
+   * Implementation of action of Frozen Lake.
+   */
+  class Action
+  {
+   public:
+    enum actions
+    {
+      Up, Down, Left, Right
+    };
+    // To store the action.
+    Action::actions action;
+
+    // Track the size of the action space.
+    static const size_t size = 4;
+  };
+
+  //! Constructor.
+  FrozenLake(size_t const& maxSteps=200,
+              size_t const& nRows=4,
+              size_t const& nCols=4,
+              double const& platformRate=0.8) : 
+    maxSteps(maxSteps),
+    nRows(nRows),
+    nCols(nCols)
+  { 
+    // Preprocess platform rate
+    this->platformRate = std::min(platformRate, 1.0);
+    this->platformRate = std::max(platformRate, 0.0);
+  }
+
+  //  Step function
+  double Sample(const State& state,
+                const Action& action,
+                State& nextState)
+  {
+    // Update the number of steps performed.
+    stepsPerformed++;
+
+    size_t newRow = state.CurRow();
+    size_t newCol = state.CurCol();
+
+    // Copy the state to the next state.
+    nextState = state;
+    
+    // Calculate new row and col.
+    if (action.action == Action::actions::Left)
+      newCol = std::max(nextState.CurCol() - 1, (unsigned long) 0);
+    else if (action.action == Action::actions::Down)
+      newRow = std::min(nextState.CurRow() + 1, nRows - 1);
+    else if (action.action == Action::actions::Right)
+      newCol = std::min(nextState.CurCol() + 1, nCols - 1);
+    else if (action.action == Action::actions::Up)
+      newRow = std::max(nextState.CurRow() - 1, (unsigned long) 0);
+    
+    // Update states by setting new current row and column, 
+    //    and manually set the description of board.
```

Comment:
```suggestion
    // and manually set the description of board.
```

Remove some extra whitespace.